src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -888,221 +1201,431 @@ public void Remove()
+            });
+
+            if (sourceIndex == 0 && length == array.Length)
             {
-                var target = new int[0];
-                Assert.Throws<NullReferenceException>(() => s_emptyDefault.CopyTo(target));
+                CopyAndInvoke(destination, destinationArray =>
+                {
+                    array.CopyTo(destinationArray, destinationIndex);
+
+                    Assert.Equal(destination.Take(destinationIndex), destinationArray.Take(destinationIndex));
+                    Assert.Equal(source, destinationArray.Skip(destinationIndex).Take(array.Length));
+                    Assert.Equal(destination.Skip(destinationIndex + array.Length), destinationArray.Skip(destinationIndex + array.Length));
+                });
+
+                if (destinationIndex == 0)
+                {
+                    CopyAndInvoke(destination, destinationArray =>
+                    {
+                        array.CopyTo(destinationArray);
+
+                        Assert.Equal(source, destinationArray.Take(array.Length));
+                        Assert.Equal(destination.Skip(array.Length), destinationArray.Skip(array.Length));
+                    });
+                }
             }
         }
 
-        [Fact]
-        public void CopyToArrayInt()
+        private static void CopyAndInvoke<T>(IEnumerable<T> source, Action<T[]> action) => action(source.ToArray());
+
+        public static IEnumerable<object[]> CopyToData()
         {
-            var source = ImmutableArray.Create(1, 2, 3);
-            var target = new int[4];
-            source.CopyTo(target, 1);
-            Assert.Equal(new[] { 0, 1, 2, 3 }, target);
+            yield return new object[] { s_manyElements, 0, new int[3], 0, 3 };
+            yield return new object[] { new[] { 1, 2, 3 }, 0, new int[4], 1, 3 };
+            yield return new object[] { new[] { 1, 2, 3 }, 0, Enumerable.Range(1, 4), 1, 3 };
+            yield return new object[] { new[] { 1, 2, 3 }, 1, new int[4], 3, 1 };
+            yield return new object[] { new[] { 1, 2, 3 }, 1, Enumerable.Range(1, 4), 3, 1 };
+        }
 
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.CopyTo(target, 0));
-            Assert.Throws<ArgumentNullException>("dest", () => source.CopyTo(null, 0));
-            Assert.Throws<ArgumentOutOfRangeException>("dstIndex", () => source.CopyTo(target, -1));
-            Assert.Throws<ArgumentException>("", () => source.CopyTo(target, 2));
+        [Theory]
+        [MemberData(nameof(Int32EnumerableData))]
+        public void CopyToInvalid(IEnumerable<int> source)
+        {
+            var array = source.ToImmutableArray();
+
+            // ImmutableArray<T>.CopyTo defers to Array.Copy for argument validation, so
+            // the parameter names here come from Array.Copy.
+
+            // TODO: The parameter names in CoreCLR need to be better.
+
+            Assert.Throws<ArgumentNullException>("dest", () => array.CopyTo(null));
```

Comment:
The parameter names in `Array.Copy` actually match that of `ImmutableArray.CopyTo` (`sourceIndex`, `destinationIndex`, `destination`, etc.) but the names in the exceptions do not match up. I will look into changing the names in CoreCLR in a later PR.